### ISSUE TITLE
Service NodePorts can be accessed via LB IPs (iptables, at least)

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#111863]
Create a single-port Service of type LB.  It gets nodeport 31234.

Wait for an LB IP to be assigned.

Access that LB IP on 31234 - it works as long as firewalls don't stop it.

This is because the NodPorts chain doesn't necessarily check IPs (hence why it is the last chain).  It finds port 31234 and forwards it to the Service.

I think we need to add rules to "KUBE-PROXY-FIREWALL" to allow valid IP+port combos and then drop other uses of those IPs.  Or something.

@danwinship since you are elbow-deep in that area.

**Comment by uablrek** [comment]
This problem does not exist for proxy-mode=ipvs.
```
vm-201 ~ # curl http://192.168.1.2:30018 > /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   232  100   232    0     0  36907      0 --:--:-- --:--:-- --:--:-- 38666
vm-201 ~ # curl http://10.0.0.80:30018  > /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to 10.0.0.80 port 30018 after 2 ms: Connection refused
```
192.168.1.2 is a node address and 10.0.0.80 is a LB-address.

**Comment by uablrek** [comment]
Please consider nodes with multiple interfaces so that access via NodePort is not restricted to the addresses in the K8s node object for instance. We sometimes use K8s on a "backend" network to isolate cluster traffic and external access from a "frontend" network;

<img src="https://raw.githubusercontent.com/Nordix/xcluster/master/ovl/network-topology/backend.svg" width="80%" />

Now, we don't use NodePort afaik but others might.

**Comment by aojea** [comment]
this one? https://github.com/kubernetes/kubernetes/pull/110289

**Comment by danwinship** [comment]
So, we _can_ work around this in kube-proxy, but I believe this bug is a side effect of GCE stuff which is not part of core Kubernetes. The rule that iptables kube-proxy creates to jump to the nodeports chain is:

    -A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain"
        -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS

`--dst-type LOCAL` means it will only match for packets that are addressed to an IP which the routing table considers to be local (meaning either the IP is assigned to a local interface, or there is an `ip route` rule intercepting traffic for that IP and delivering it to a local interface). Neither iptables kube-proxy nor the GCP CloudProvider adds LB IPs as local IPs, or adds routes directing LB IP traffic to the local host, so this rule _should not match_ LB IP traffic unless something besides core Kubernetes is mucking with the routing of LB IP traffic.

I know in OpenShift-on-GCP, there _is_ a service that does special handling for LoadBalancer IPs, and I'm pretty sure it's a fork of something that runs on GCE. The behavior in OpenShift has changed over the years, and it now uses iptables rules rather than routes, but it ends up having [the same bug](https://bugzilla.redhat.com/show_bug.cgi?id=2094408) (which I had assumed GCE _didn't_ have), and my conclusion there was that we shouldn't be running that script for LB IPs that correspond to Kubernetes Services; it should only be used for LB IPs that _aren't_ being managed by kube-proxy. (Of course if there are LB IPs that are used for both kube-proxy-managed Services and for other things, then that rule wouldn't work...)

Assuming I'm right and this service does exist in GCE, we should probably document its existence and behavior better in the kube-proxy code/docs...

**Comment by aojea** [comment]
note: all implementations that add IP addresses to the node may be affected by this, per example, platforms that use Keepalived on nodes for ExternalIPs  or LoadBalancer IPs

**Comment by thockin** [comment]
Yeah, any LB implementation that is VIP-ish should have this problem - something needs to accept VIP traffic locally.

My fear is that we end up with 1 "ALLOW" `filter` rule for every service-ip-port and then final "DROP" rules, increasing the number of iptables rules.  With things like `externalIPs` it's not even obvious when to put that rule, since a single externalIP could be used on multiple Services (with disjoint sets of ports).  Also, annoyingly, this speaks to the value of KUBE-MARK-DROP, since we already enumerate all service-ip-ports in `nat`.

**Comment by danwinship** [comment]
> My fear is that we end up with 1 "ALLOW" filter rule for every service-ip-port and then final "DROP" rules, increasing the number of iptables rules

Tim mentioned briefly in the meeting that we have all these problems with bad Service defaults, and I feel like part of the answer here might be to encourage people to use `--nodeport-addresses` or something similar; then we can just change the `KUBE-NODEPORTS` jump rule to only match traffic on those IPs, rather than having to un-match everything else.

(Also, if people use a `--nodeport-addresses` that doesn't include 127.0.0.1 then we can also get rid of the localhost-nodeport junk.)

But even without that, we still don't need 1 rule per IP; we can just make the iptables proxy use ipsets like the ipvs proxy does... there's no real reason other than "history" for why it doesn't.

**Comment by danwinship** [comment]
> encourage people to use `--nodeport-addresses` or something similar

eg
`klog.Warning("--nodeport-addresses not specified; NodePorts will be available on all local IPs!")`

**Comment by aojea** [comment]
yeah, the `Access that LB IP on 31234 - it works as long as firewalls don't stop it.`  is totally valid if the LB IP belong to the Node ... but is this new? this should be always working like this

**Comment by danwinship** [comment]
Yes, it always worked like this, but it was never _intended_ to work like this.

**Comment by danwinship** [comment]
If we added `--nodeport-addresses node-ip` meaning "only accept NodePort connections on the node's primary IP (of the relevant IP family)", would that work with the various cloud load balancers that actually use NodePorts? If not, would it work if we had it mean "only accept NodePort connections on IPs in `node.Status.Addresses`"? (Yes, in both cases it would imply "no NodePorts on localhost; people who want localhost NodePorts would have to continue not using `--nodeport-addresses` or else specify the right IP/CIDR by hand.)

**Comment by danwinship** [comment]
#123105 implements the suggestion above, of adding a new option to request that NodePorts only work on the node's primary IP, and it adds a warning if you don't use that option and don't manually set `--nodeport-addresses`.

Additionally, the nftables backend implements that functionality by default (#122724) and we should update the `v1alpha2.KubeProxyConfiguration` plan (https://github.com/kubernetes/enhancements/pull/4337) to enable this by default as well.
